## 3,1 다시 보는 초난감 DAO

### 3.1.1 예외처리 기능을 갖춘 DAO

DB 커넥션이라는 제한적인 리소스를 공유하여 사용하는 JDBC 코드는 흐름을 벗어나는 경우에도 사용한 리소스를 반드시 반환하도록 예외처리를 해줘야 한다. 그렇지 않으면 시스템에 심각한 문제가 발생할 수 있다.

### JDBC 수정 기능의 예외처리 코드

```java
public void deleteAll() throws SQLException {
    Connection c = dataSource.getConnection();
    
    PreparedStatement ps = c.prepareStatement("delete from users");
    ps.executeUpdate();
    
    ps.close();
    c.close();
}
```

이 메소드에서는 Connectino, PreparedStatement라는 두 개의 공유 리소스를 가져와서 사용한다. 하지만 만약 공유 리소스를 처리하는 중 예외가 발생하면 해당 메소드를 빠져나가게 된다. 이때 문제는 close() 메소드가 실행되지 않아서 `사용한 리소스가 제대로 반환되지 않을 수 있다는 점이다.`

일반적으로 서버에서는 제한된 개수의 DB 커넥션을 만들어서 재사용 가능한 풀로 관리한다. 하지만 이런 식으로 오류가 발생할 때 마다 `반환되지 못한 Connection이 계속 쌓이게 된다면 리소스가 모자라다는 심각한 오류가 발생하며 서버가 중단될 수 있다.`

> **리소스 반환과 close()**
> <br />
>Connection과 PreparedStatement는 보통 풀 방식으로 운영된다. 미리 정해진 풀 안에 제한된 수의 리소스를 만들어 두고 필요할 때 이를 할당하고, 반환하면 다시 풀에 넣는 방식으로 운영된다. 요청이 매우 많은 서버환경에서는 이렇게 미리 풀에 리소스를 만들어두어 사용하는 편이 유리하다. 대신, 사용한 리소스는 빠르게 반환해야 한다. 그렇지 않으면 풀에 있는 리소스가 고갈되고 결국 문제가 발생한다. close() 메소드는 사용한 리소스를 풀로 다시 돌려주는 역할을 한다.
 
그래서 이런 JDBC 코드에서는 어떤 상황에서도 가져온 리소스를 반환하도록 try/catch/finally 구문 사용을 권장하고 있다.

```java
  public void deleteAll() throws SQLException {
    Connection c = null;
    PreparedStatement ps = null;
    try {
      c = dataSource.getConnection();
      ps = c.prepareStatement("delete from users");
      ps.executeUpdate();
    } catch (SQLException e) {
      throw e;
    } finally {
      if (ps != null) {
        try {
          ps.close();
        } catch (SQLException e) {
        }
      }
      if (c != null) {
        try {
          c.close();
        } catch (SQLException e) {
        }
      }
    }
  }
```

### JDBC 조회 기능의 예외처리

조회를 위한 JDBC 코드는 ResultSet이 포함되기 때문에 좀 더 복잡해진다.

```java
  public int getCount() throws SQLException {
    Connection c= null;
    PreparedStatement ps = null;
    ResultSet rs = null;
    
    try {
      c = dataSource.getConnection();
      ps = c.prepareStatement("select count(*) from users");
      rs = ps.executeQuery();

      rs.next();
      return rs.getInt(1);
    } catch (SQLException e) {
      throw e;
    } finally {
      if (ps != null) {
        try {
          ps.close();
        } catch (SQLException e) {
        }
      }
      if (c != null) {
        try {
          c.close();
        } catch (SQLException e) {
        }
      }
      if (rs != null) {
        try {
          rs.close();
        } catch (SQLException e) {
          
        }
      }
    }
  }
```

이제 서버환경에서도 안정적으로 수행될 수 있는 DAO가 완성됐다.